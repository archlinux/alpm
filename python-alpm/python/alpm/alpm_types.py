# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from typing import Union

from typing_extensions import TypeAlias

from .alpm import alpm_types as _mod

Checksum: TypeAlias = Union[
    _mod.Blake2b512Checksum,
    _mod.Md5Checksum,
    _mod.Sha1Checksum,
    _mod.Sha224Checksum,
    _mod.Sha256Checksum,
    _mod.Sha384Checksum,
    _mod.Sha512Checksum,
]

__all__ = [*_mod.__all__, "Checksum"]
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
