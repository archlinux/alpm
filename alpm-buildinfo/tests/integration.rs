use std::{str::FromStr, thread};

use alpm_buildinfo::{BuildInfoV1, BuildInfoV2, Schema};
use alpm_types::SchemaVersion;
use assert_cmd::Command;
use insta::assert_snapshot;
use rstest::rstest;
use testdir::testdir;
use testresult::TestResult;

pub const VALID_BUILDINFO_V1_DATA: &str = r#"
builddate = 1
builddir = /build
buildenv = envfoo
buildenv = envbar
format = 1
installed = bar-1.2.3-1-any
installed = beh-2.2.3-4-any
options = some_option
options = !other_option
packager = Foobar McFooface <foobar@mcfooface.org>
pkgarch = any
pkgbase = foo
pkgbuild_sha256sum = b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
pkgname = foo
pkgver = 1:1.0.0-1
"#;

pub const VALID_BUILDINFO_V2_DATA: &str = r#"
builddate = 1
builddir = /build
startdir = /startdir/
buildtool = devtools
buildtoolver = 1:1.2.1-1-any
buildenv = envfoo
buildenv = envbar
format = 2
installed = bar-1.2.3-1-any
installed = beh-2.2.3-4-any
options = some_option
options = !other_option
packager = Foobar McFooface <foobar@mcfooface.org>
pkgarch = any
pkgbase = foo
pkgbuild_sha256sum = b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
pkgname = foo
pkgver = 1:1.0.0-1
"#;

#[derive(Default)]
pub struct BuildInfoInput {
    pub format: Schema,
    pub builddate: Option<String>,
    pub builddir: Option<String>,
    pub buildenv: Option<Vec<String>>,
    pub installed: Option<Vec<String>>,
    pub options: Option<Vec<String>>,
    pub packager: Option<String>,
    pub pkgarch: Option<String>,
    pub pkgbase: Option<String>,
    pub pkgbuild_sha256sum: Option<String>,
    pub pkgname: Option<String>,
    pub pkgver: Option<String>,

    // V2 fields
    pub startdir: Option<String>,
    pub buildtool: Option<String>,
    pub buildtoolver: Option<String>,
}

/// Validate the V1 schema.
/// The version is automatically determined from the file
#[test]
fn validate_valid_buildinfov1() -> TestResult {
    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.arg("validate");
    cmd.write_stdin(VALID_BUILDINFO_V1_DATA);
    cmd.assert().success();
    Ok(())
}

/// Validate the V2 schema.
/// The version is automatically determined from the file
#[test]
fn validate_valid_buildinfov2() -> TestResult {
    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.arg("validate");
    cmd.write_stdin(VALID_BUILDINFO_V2_DATA);
    cmd.assert().success();
    Ok(())
}

/// Force a v2 validation on a v1 buildinfo
#[test]
fn wrong_schema_buildinfov1_as_v2() -> TestResult {
    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.args(["validate", "-s", "2"]);
    cmd.write_stdin(VALID_BUILDINFO_V1_DATA);
    cmd.assert().failure().code(1);
    Ok(())
}

/// Force a v1 validation on a v2 buildinfo
#[test]
fn wrong_schema_buildinfov2_as_v1() -> TestResult {
    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.args(["validate", "-s", "1"]);
    cmd.write_stdin(VALID_BUILDINFO_V2_DATA);
    cmd.assert().failure().code(1);
    Ok(())
}

/// Format BUILDINFO as JSON.
#[rstest]
#[case::buildinfov1_as_json(VALID_BUILDINFO_V1_DATA)]
#[case::buildinfov2_as_json(VALID_BUILDINFO_V2_DATA)]
fn format_buildinfo_and_serialize_as_json(#[case] data: &str) -> TestResult {
    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.args(["format", "-p"]);
    cmd.write_stdin(data);
    let cmd = cmd.unwrap();
    let build_info = String::from_utf8_lossy(&cmd.stdout);
    assert_snapshot!(
        thread::current().name().unwrap().to_string(),
        build_info.to_string()
    );
    Ok(())
}

#[rstest]
#[case::buildinfov1_all_fields(
    BuildInfoInput {
        format: Schema::default(),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: Some(vec!["foo".to_string(), "bar".to_string()]),
        installed: Some(vec!["bar-1.2.3-1-any".to_string(), "beh-2.2.3-4-any".to_string()]),
        options: Some(vec!["some_option".to_string(), "!other_option".to_string()]),
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: None,
        buildtool: None,
        buildtoolver: None,
    },
)]
#[case::buildinfov1_optional_fields(
    BuildInfoInput {
        format: Schema::default(),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: None,
        installed: None,
        options: None,
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: None,
        buildtool: None,
        buildtoolver: None,
    },
)]
#[case::buildinfov2_all_fields(
    BuildInfoInput {
        format: Schema::V2(SchemaVersion::from_str("2").unwrap()),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: Some(vec!["foo".to_string(), "bar".to_string()]),
        installed: Some(vec!["bar-1.2.3-1-any".to_string(), "beh-2.2.3-4-any".to_string()]),
        options: Some(vec!["some_option".to_string(), "!other_option".to_string()]),
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: Some("/startdir/".to_string()),
        buildtool: Some("devtools".to_string()),
        buildtoolver: Some("1:1.2.1-1-any".to_string()),
    },
)]
#[case::buildinfov2_optional_fields(
    BuildInfoInput {
        format: Schema::V2(SchemaVersion::from_str("2").unwrap()),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: None,
        installed: None,
        options: None,
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: Some("/startdir/".to_string()),
        buildtool: Some("devtools".to_string()),
        buildtoolver: Some("1:1.2.1-1-any".to_string()),
    },
)]
fn write_buildinfo_via_env(#[case] buildinfo_input: BuildInfoInput) -> TestResult {
    test_write_buildinfo(buildinfo_input, true)
}

#[rstest]
#[case::buildinfov1_all_fields(
    BuildInfoInput {
        format: Schema::default(),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: Some(vec!["foo".to_string(), "bar".to_string()]),
        installed: Some(vec!["bar-1.2.3-1-any".to_string(), "beh-2.2.3-4-any".to_string()]),
        options: Some(vec!["some_option".to_string(), "!other_option".to_string()]),
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: None,
        buildtool: None,
        buildtoolver: None,
    },
)]
#[case::buildinfov1_optional_fields(
    BuildInfoInput {
        format: Schema::default(),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: None,
        installed: None,
        options: None,
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: None,
        buildtool: None,
        buildtoolver: None,
    },
)]
#[case::buildinfov2_all_fields(
    BuildInfoInput {
        format: Schema::V2(SchemaVersion::from_str("2").unwrap()),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: Some(vec!["foo".to_string(), "bar".to_string()]),
        installed: Some(vec!["bar-1.2.3-1-any".to_string(), "beh-2.2.3-4-any".to_string()]),
        options: Some(vec!["some_option".to_string(), "!other_option".to_string()]),
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: Some("/startdir/".to_string()),
        buildtool: Some("devtools".to_string()),
        buildtoolver: Some("1:1.2.1-1-any".to_string()),
    },
)]
#[case::buildinfov2_optional_fields(
    BuildInfoInput {
        format: Schema::V2(SchemaVersion::from_str("2").unwrap()),
        builddate: Some("1".to_string()),
        builddir: Some("/build".to_string()),
        buildenv: None,
        installed: None,
        options: None,
        packager: Some("Foobar McFooface <foobar@mcfooface.org>".to_string()),
        pkgarch: Some("any".to_string()),
        pkgbase: Some("foo".to_string()),
        pkgbuild_sha256sum: Some("b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c".to_string()),
        pkgname: Some("foo".to_string()),
        pkgver: Some("1:1.0.0-1".to_string()),
        startdir: Some("/startdir/".to_string()),
        buildtool: Some("devtools".to_string()),
        buildtoolver: Some("1:1.2.1-1-any".to_string()),
    },
)]
fn write_buildinfo_via_cli(#[case] buildinfo_input: BuildInfoInput) -> TestResult {
    test_write_buildinfo(buildinfo_input, true)
}

/// Test writing a buildinfo file either via CLI or environment variables.
fn test_write_buildinfo(buildinfo_input: BuildInfoInput, use_env: bool) -> TestResult {
    let dir = testdir!();
    let test_name = thread::current().name().unwrap().to_string();

    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.args(["create".to_string(), format!("v{}", buildinfo_input.format)])
        .current_dir(dir.clone());
    if use_env {
        set_buildinfo_env(&mut cmd, &buildinfo_input);
    } else {
        set_buildinfo_args(&mut cmd, &buildinfo_input);
    }
    cmd.assert().success();
    let file = dir.join(".BUILDINFO");
    assert!(file.exists());

    let contents = std::fs::read_to_string(&file)?;
    let build_info = match buildinfo_input.format {
        Schema::V1(_) => BuildInfoV1::from_str(&contents)?.to_string(),
        Schema::V2(_) => BuildInfoV2::from_str(&contents)?.to_string(),
        _ => unimplemented!("Unimplemented schema!"),
    };
    assert_snapshot!(test_name, build_info.to_string());

    let mut cmd = Command::cargo_bin("alpm-buildinfo")?;
    cmd.args([
        "validate".to_string(),
        "-s".to_string(),
        buildinfo_input.format.to_string(),
        file.to_string_lossy().to_string(),
    ]);
    cmd.assert().success();

    Ok(())
}

fn set_buildinfo_args(cmd: &mut Command, input: &BuildInfoInput) {
    if let Some(ref builddate) = input.builddate {
        cmd.args(["--builddate", builddate]);
    }
    if let Some(ref builddir) = input.builddir {
        cmd.args(["--builddir", builddir]);
    }
    if let Some(ref buildenv) = input.buildenv {
        for env in buildenv.iter() {
            cmd.args(["--buildenv", env]);
        }
    }
    if let Some(ref installed) = input.installed {
        for package in installed.iter() {
            cmd.args(["--installed", package]);
        }
    }
    if let Some(ref options) = input.options {
        for option in options.iter() {
            cmd.args(["--options", option]);
        }
    }
    if let Some(ref packager) = input.packager {
        cmd.args(["--packager", packager]);
    }
    if let Some(ref pkgarch) = input.pkgarch {
        cmd.args(["--pkgarch", pkgarch]);
    }
    if let Some(ref pkgbase) = input.pkgbase {
        cmd.args(["--pkgbase", pkgbase]);
    }
    if let Some(ref pkgbuild_sha256sum) = input.pkgbuild_sha256sum {
        cmd.args(["--pkgbuild-sha256sum", pkgbuild_sha256sum]);
    }
    if let Some(ref pkgname) = input.pkgname {
        cmd.args(["--pkgname", pkgname]);
    }
    if let Some(ref pkgver) = input.pkgver {
        cmd.args(["--pkgver", pkgver]);
    }

    if let Schema::V2(_) = input.format {
        if let Some(ref startdir) = input.startdir {
            cmd.args(["--startdir", startdir]);
        }
        if let Some(ref buildtool) = input.buildtool {
            cmd.args(["--buildtool", buildtool]);
        }
        if let Some(ref buildtoolver) = input.buildtoolver {
            cmd.args(["--buildtoolver", buildtoolver]);
        }
    }
}

fn set_buildinfo_env(cmd: &mut Command, input: &BuildInfoInput) {
    if let Some(ref builddate) = input.builddate {
        cmd.env("BUILDINFO_BUILDDATE", builddate);
    }
    if let Some(ref builddir) = input.builddir {
        cmd.env("BUILDINFO_BUILDDIR", builddir);
    }
    if let Some(ref buildenv) = input.buildenv {
        cmd.env("BUILDINFO_BUILDENV", buildenv.join(" "));
    }
    if let Some(ref installed) = input.installed {
        cmd.env("BUILDINFO_INSTALLED", installed.join(" "));
    }
    if let Some(ref options) = input.options {
        cmd.env("BUILDINFO_OPTIONS", options.join(" "));
    }
    if let Some(ref packager) = input.packager {
        cmd.env("BUILDINFO_PACKAGER", packager);
    }
    if let Some(ref pkgarch) = input.pkgarch {
        cmd.env("BUILDINFO_PKGARCH", pkgarch);
    }
    if let Some(ref pkgbase) = input.pkgbase {
        cmd.env("BUILDINFO_PKGBASE", pkgbase);
    }
    if let Some(ref pkgbuild_sha256sum) = input.pkgbuild_sha256sum {
        cmd.env("BUILDINFO_PKGBUILD_SHA256SUM", pkgbuild_sha256sum);
    }
    if let Some(ref pkgname) = input.pkgname {
        cmd.env("BUILDINFO_PKGNAME", pkgname);
    }
    if let Some(ref pkgver) = input.pkgver {
        cmd.env("BUILDINFO_PKGVER", pkgver);
    }

    if let Schema::V2(_) = input.format {
        if let Some(ref startdir) = input.startdir {
            cmd.env("BUILDINFO_STARTDIR", startdir);
        }
        if let Some(ref buildtool) = input.buildtool {
            cmd.env("BUILDINFO_BUILDTOOL", buildtool);
        }
        if let Some(ref buildtoolver) = input.buildtoolver {
            cmd.env("BUILDINFO_BUILDTOOLVER", buildtoolver);
        }
    }
}
