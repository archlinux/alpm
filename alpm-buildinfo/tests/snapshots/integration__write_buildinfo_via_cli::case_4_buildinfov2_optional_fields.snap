---
source: alpm-buildinfo/tests/integration.rs
expression: build_info.to_string()
snapshot_kind: text
---
format = 2
pkgname = foo
pkgbase = foo
pkgver = 1:1.0.0-1
pkgarch = any
pkgbuild_sha256sum = b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
packager = Foobar McFooface <foobar@mcfooface.org>
builddate = 1
builddir = /build
startdir = /startdir/
buildtool = devtools
buildtoolver = 1:1.2.1-1-any
